# Instructions on released search space

We released the search space for 21 cancers [here](https://genome.med.nyu.edu/public/yarmarkovichlab/ImmunoVerse/search_space/). The 
intended usage is when the users have in-house immunopeptidome data, such as pancreatic cancer without
matched RNA or genomics data, the `PAAD` search space can be readily
used for immunopeptidome search to find more non-canonical tumor specific antigens

In each folder, fasta files generally include:

1. `ensembl_proteins.fasta`: This is the canonical proteins normally used for immunopeptidome search,
we curate it from Ensembl to have nicely-linked ENST and ENSG ID.
2. `uniprot_reviewed_curated_addition.fasta`: This includes reviewed human isoforms downloaded from Uniprot
and carefully compared with our Ensembl canonical proteome. Including this database is useful, because if a peptide
maps to both a fusion transcript and a reviewed isoform, we know the peptide is not from the fusion region but the exonic one.
3. `splicing.fasta`: Splicing flanking sequence after in silico translation
4. `nuorf.fasta`: We manually curated the whole [nuORF database](https://pubmed.ncbi.nlm.nih.gov/34663921/) and derived the actual peptide sequences which were not available 
in original paper.
5. `mutation.fasta`: mutation and flanking sequence
6. `intron.fasta`: the intron transcirpt and peptides
7. `fusion.fasta`: the fusion transcript and peptides
8. `TE_self_translate.fasta`: self-translated TE element
9. `various uniprot microbiome proteome`: pathogen proteome


# TE explanation of headers

`MER5B_dup847|chr1:148430344-148430511:-|1.249905904261159|0.0440972821239617|4.82|151.0|1|1|sense`: 
locus annotation is from TEtranscript, hg38 coordinates and strand can be queried against UCSC genome browser,
`1.24` is the median cpm in tumor, `0.04` is mean cpm in normal, `4.82` is log fold change, `151` is the number of 
samples with this TE expression, first `1` means we translate it in the frame 1, meaning starts the translation from the 1st
nucleotide, second `1` is less important, it means it's the first peptide segment delimited by stop codons. `sense` means
the peptide is translated in the sense orientation.

## Intron explanation of headers
`ENSG00000000938.13,FGR,chr1,-,I1.1,27626570,27635064|52|17.42|1|intron`: 
`52` means 52 samples have this intron retention, `17.42` means the median exon coverage (TPM) in tumor,
`1` is less important, it means when a intron-retention results in more than one possible open-reading-frames,
we recorded all, and this is the first ORF.

## Mutation explanation of headers
`NPM1|p.W288Cfs*12|9|0.37|ENSG00000181163|chr5:171410539-171410540|-/TCTG|frameshift_variant`:
`9` means the number of patients, `0.37` is the median clonality.

## Splicing explanation of headers
`chr10:100176070-100179201|104|17.57|0.12|7.17|-|0|ERLIN1,ERLIN1|3`: `104` is the
number of patients, `17.57` is the mean read count in tumor, `0.12` is the mean 
read count in normal, `7.17` is log fold change, `0` is the phase, if the phase is 1, it means
the junction codon is generated by taking 2nt from first half and 1nt from second half. `3` means
it is the 3rd peptide segment.


